### PR TITLE
DEX-774: some matching set fundamentals

### PR DESCRIPTION
**Pull request by naknomum**
## Pull Request Overview

- Modifies identification process to use Elasticsearch query to limit Annotations which will be matched against (_"matching set"_)
- Introduces a _**default matching set**_ when no query is provided
- Provides some API endpoints regarding matching sets
---

**Notes**

Querying to produce a matching set is simply a variation on a generic Elasticsearch query against the Annotation index.  This query is slightly modified to add some required limitations, such as not being in the same Encounter as the query Annotation.

**Using an Elasticsearch query in ID pipeline**

An additional field `matching_set` can be added to each section of `ia_config`, such as:
```
"ia_config": [
        {
            "algorithms": ["hotspotter_nosv"],
            "matching_set": {"bool": {"filter": ................ }}
        }
]
```
If there is no `matching_set` provided, the _default_ one will be used.

**REST API Examples**

Some access to matching-set data is available via new endpoints, as explained below.

This GET will return the **default matching set Elasticsearch _query_** for a given Annotation:
```
GET /api/v1/annotations/matching_set/query/ANNOT_GUID

{
    "bool": {
        "filter": [
            {
                "exists": {
                    "field": "content_guid"
                }
            },
            {
                "bool": {
                    "minimum_should_match": 1,
                    "should": [
                        {
                            "term": {
                                "viewpoint": "upbackleft"
                            }
                        },
   ...
```
POST to the same endpoint with a request body of a _custom_ Elasticsearch query will return the actual query that will be used to find the matching set (which may be modified slightly with some required additions).
```
POST /api/v1/annotations/matching_set/query/ANNOT_GUID
    { ... custom Elasticsearch query in body ...}

{
    "bool": {
        "filter": 
            ... query as it will be used ...
   ...
```

These will return the actual resultant matching set contents, as only the GUIDs of the set's Annotations.  It can be useful to retrieve the size of the set the Annotation will be matched against.
The GET variation uses the _default matching set_, and the POST will apply the _custom_ query passed via the request.

```
GET /api/v1/annotations/matching_set/ANNOT_GUID

[
    "105a5246-9876-4f66-88dd-d7f2c2c7f6db",
    "765107f7-0047-4e8f-845c-124dc7ffd8ed",
    "8dfce746-2f3a-4f96-a802-1044b90a0bf4"
]
```

### CODE REVIEW

File: app/modules/annotations/models.py [hunk 1]
```diff
@@ -187,6 +185,97 @@ def get_asset_src(self):
...
+    # and may not be ideal for "general usage" cases
+    def get_related_extended_data(self):
+        if hasattr(self, '_enc'):
+            return self._enc, self._enc_edm, self._sight_edm
+        if not self.encounter_guid:
+            return None, None, None
+        from app.modules.encounters.models import Encounter
+        from flask import current_app
+
+        self._enc = Encounter.query.get(self.encounter_guid)
```
Comment (hwindsor) [inline app/modules/annotations/models.py:253]: Set this last. If any of the EDM gets fail, you will be left with this set but no EDM data

File: app/modules/annotations/schemas.py [hunk 2]
```diff
@@ -56,3 +56,80 @@ class Meta(BaseAnnotationSchema.Meta):
             Annotation.updated.key,
             'asset_src',
         )
+
```
Comment (hwindsor) [inline app/modules/annotations/schemas.py:59]: Why are these methods here and not on the annot model itself?
Note I'm thinking of moving the cache to the different classes so that these caches can be reused

File: app/modules/annotations/schemas.py [hunk 3]
```diff
@@ -56,3 +56,80 @@ class Meta(BaseAnnotationSchema.Meta):
...
+    Schema for indexing by Elasticsearch
+
+    Note: can be expensive (as it delves into related objects as well as EDM), so best not to use
+    for purposes other than ES indexing.
+    """
+
+    keywords = base_fields.Function(get_keywords_flat)
+    locationId = base_fields.Function(get_locationId)
+    taxonomy_guid = base_fields.Function(get_taxonomy_guid)
+    owner_guid = base_fields.Function(get_owner_guid)
```
Comment (hwindsor) [inline app/modules/annotations/schemas.py:113]: Why have owner guid? All other APIs have the owner PublicUserSchema.

File: app/modules/annotations/schemas.py (hunk 2 above)
Comment (naknomum) [reply app/modules/annotations/schemas.py:59]: simply put, i did not want to make these methods which access properties _from the edm_ publicly usable (yet?).  my guess is they would tend to be expensive to use -- even with my vague attempt at caching edm values.   if/when something needs to use them, i would be fine with (cautiously) moving them to the actual Annotation model -- but perhaps with a bit more consideration of cost to get the values.   (and once edm goes away, this would be a moot point and i would be all for having them on the model.)

so since -- for now -- the only place that needs to get this is for this schema; i just kind of plunked them here "for internal use only" sorta thing.  (knowing the cost was worth bearing for when it needed to be indexed -- which hopefully will ultimately be done as minimal as possible.)

File: app/modules/annotations/schemas.py (hunk 3 above)
Comment (naknomum) [reply app/modules/annotations/schemas.py:113]: this schema is for **elasticsearch indexing** only -- so it is what we want to search against -- not something that is used like a normal api. e.g. one would want to simply match the owner _guid_, not the details about that User.   similarly, i made the **keywords** just the list of text-values basically -- with the assumption that would be enough to search against (via Elasticsearch).  however, i could see modifying this one to _also_ allow searching on the keyword guid as well.

in short, these elastic-search-specific schema are what the ES _document_ ends up looking like.

File: app/modules/annotations/schemas.py (hunk 2 above)
Comment (hwindsor) [reply app/modules/annotations/schemas.py:59]: I see the reasoning (for the first two methods but not the latter six, unless I'm mistaken they don't use EDM)
Regarding the implementation, I suspect I will need to do something equivalent for individual, Encounter and Sighting so I think get_locationId_from_edm on the encounter model may be better.

File: app/modules/annotations/schemas.py (hunk 3 above)
Comment (naknomum) [reply app/modules/annotations/schemas.py]: addressing in refactored code

File: app/modules/annotations/schemas.py (hunk 2 above)
Comment (naknomum) [reply app/modules/annotations/schemas.py:59]: addressing in refactored code

File: app/modules/annotations/schemas.py [hunk 4]
```diff
@@ -56,3 +56,62 @@ class Meta(BaseAnnotationSchema.Meta):
...
         )
+
+
+def get_keyword_values(annot):
+    if not annot.keyword_refs:
+        return []
+    return sorted([ref.keyword.value for ref in annot.keyword_refs])
+
+
+# these just handle string conversion basically
```
Comment (hwindsor) [inline app/modules/annotations/schemas.py:67]: I did mean that these should be class methods

File: app/modules/annotations/models.py [hunk 5]
```diff
@@ -187,6 +185,96 @@ def get_asset_src(self):
...
+            raise ValueError(f'{self} has no Encounter so cannot be matched against')
+        if not query or not isinstance(query, dict):
+            query = self.get_matching_set_default_query()
+        log.info(f'finding matching set for {self} using query {query}')
+        return self.elasticsearch(query, load=load)
+
+    def get_matching_set_default_query(self):
+        # n.b. default will not take any locationId or ownership into consideration
+        parts = {'filter': []}
+        enc, enc_edm, sight_edm = self.get_related_extended_data()
```
Comment (hwindsor) [inline app/modules/annotations/models.py:199]: hasn't this function gone now?

File: app/extensions/__init__.py [hunk 6]
```diff
@@ -482,6 +483,27 @@ def current_user_has_edit_permission(self):
...
+    def get_edm_complete_data(self, use_cache=True):
+        from app.extensions import is_extension_enabled
+        from flask import current_app
+
+        # this will prevent HoustonModel objects from using this
+        if FeatherModel not in self.__class__.__bases__:
+            raise NotImplementedError('only available on FeatherModels')
+        if not is_extension_enabled('edm'):
+            return None
+        edm_data = current_app.edm.get_dict(
```
Comment (hwindsor) [inline app/extensions/__init__.py:498]: where's the caching?

File: app/modules/annotations/models.py (hunk 5 above)
Comment (naknomum) [reply app/modules/annotations/models.py:199]: work in progress.  same with caching.  :)   as the commit message says:
> new way to get edm-data on FeatherModel objects (currently cache-less)

File: app/extensions/__init__.py [hunk 7]
```diff
@@ -470,6 +471,38 @@ def current_user_has_edit_permission(self):
...
+        if FeatherModel not in self.__class__.__bases__:
+            raise NotImplementedError('only available on FeatherModels')
+        if not is_extension_enabled('edm'):
+            return None
+        # going to give cache a life of 5 min kinda arbitrarily
+        time_now = int(time.time())
+        if (
+            use_cache
+            and hasattr(self, '_edm_cached_data')
+            and time_now - self._edm_cached_data.get('_edm_cache_created', 0) < 300
```
Comment (hwindsor) [inline app/extensions/__init__.py:492]: I really dislike arbitrary numbers in code. Can you give this 300 a name please?
One of the guys I mentored sent me a message this morning saying he'd thought of me when creating a SecondsPerMinute constant.

File: app/extensions/__init__.py (hunk 6 above)
Comment (naknomum) [reply app/extensions/__init__.py:497]: now with caching!  :)

**Review by colinwkingen — APPROVED** [review]
LGTM
